### PR TITLE
dev-qt/qt3d: ~riscv keyworing

### PR DESCRIPTION
**Pull request by Rabenda**
dev-qt/qtgamepad: ~riscv keywording
Make repoman happy in dev-qt/qt3d

### CODE REVIEW

File: dev-qt/qt3d/qt3d-5.15.2-r11.ebuild
Comment (arthurzam) [inline dev-qt/qt3d/qt3d-5.15.2-r11.ebuild:32]: It looks like it would be a "long time" downstream patch (which is thankfully quite easy to maintain).
In that case, I think it would be better to use the `${PN}-5.15.2` instead of `${P}`, as it would make bumping easier in future.

**Review by arthurzam — COMMENTED** [review]
I'm not sure why I was called, as I'm not maintainer from Qt project, my here are my two cents from this simple PR.
But I do think more input from someone from Qt project is a good idea.

File: dev-qt/qt3d/qt3d-5.15.2-r11.ebuild
Comment (dlan17) [reply dev-qt/qt3d/qt3d-5.15.2-r11.ebuild:32]: oops, @arthurzam sorry for highlighting you.. I probably should at pesa
basically, I think it's a quite trivial patch, should be easy to maintain

anyway, thanks for the review/comments

Comment (Pesa) [reply dev-qt/qt3d/qt3d-5.15.2-r11.ebuild:32]: agree with @arthurzam's comment

**Review by Pesa — CHANGES_REQUESTED** [review]
small typo in the 2nd commit message

File: dev-qt/qt3d/files/qt3d-5.15.2-riscv-atomic.patch
Comment (Rabenda) [inline dev-qt/qt3d/files/qt3d-5.15.2-riscv-atomic.patch:10]: It seems wrong @dlan17 
qtConfig(std-atomic64) has 64bit atomic. But it should determine !qtConfig(std-atomic64) to use latomic.
Original error is `__atomic_exchange_1` missing. ATOMIC_NON64BIT is needed

Comment (Rabenda) [reply dev-qt/qt3d/files/qt3d-5.15.2-riscv-atomic.patch:10]: https://doc.qt.io/qt-5/qmake-test-function-reference.html#qtcompiletest-test

qtcompiletest can resolve it

Comment (Rabenda) [reply dev-qt/qt3d/files/qt3d-5.15.2-riscv-atomic.patch:10]: In use qtcompiletest case, it needs two test-qt-projects in patch. It looks not graceful

File: dev-qt/qt3d/qt3d-5.15.2-r11.ebuild
Comment (Rabenda) [reply dev-qt/qt3d/qt3d-5.15.2-r11.ebuild]: Now use ${PN}-5.15.2

File: dev-qt/qt3d/files/qt3d-5.15.2-riscv-atomic.patch
Comment (Rabenda) [reply dev-qt/qt3d/files/qt3d-5.15.2-riscv-atomic.patch]: New patch solve it.

File: dev-qt/qt3d/files/qt3d-5.15.2-riscv-atomic.patch
Comment (Rabenda) [reply dev-qt/qt3d/files/qt3d-5.15.2-riscv-atomic.patch:65]: Ok, I will remove it.

**Comment by Pesa** [comment]
> Please test https://invent.kde.org/qt/qt/qtbase/-/merge_requests/138 applied against qtcore alternatively.

And please @Rabenda test the above patch first. If we can solve this properly in qtcore, we will no longer have to patch a bunch of other packages.
That test is for function pointers so I'm not convinced the patch will work for 1- and 2-byte atomics (which is the problem on riscv), but it's worth trying it, and maybe we can adapt the test to check for more variations, like you did in your latest patch?

**Comment by Rabenda** [comment]
https://github.com/riscv-collab/riscv-gcc/issues/12 use Option 4
https://stackoverflow.com/questions/23250863/difference-between-pthread-and-lpthread-while-compiling
In riscv, "-pthread" equivalent to add link option "--push-state --as-needed -latomic --pop-state"

**Comment by Rabenda** [comment]
https://github.com/felixonmars/archriscv-packages/blob/master/qt5-base/riscv64.patch

Archlinux riscv fix method

**Comment by a17r** [comment]
It would be preferable to make that a patch instead, so it could theoretically at least be upstreamed.

File: dev-qt/qtcore/files/qtcore-5.15.2-atomic.patch
Comment (dlan17) [inline dev-qt/qtcore/files/qtcore-5.15.2-atomic.patch:41]: hi @Pesa @a17r , it occur to me that current '-pthread' version is much better (less invasive), what do you think?

@Rabenda can you also report this to qt upstream?

**Comment by dlan17** [comment]
hi all, please hold this for a while, we will try to rebase the patch to qt-5.15.3 (as @a17r suggested) and re-test

**Pull request closed without merging**
